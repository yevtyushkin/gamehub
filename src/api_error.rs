use serde::{Deserialize, Serialize};
use std::borrow::Cow;

/// Outbound API error.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiError<'a> {
    /// Module where the error has occurred, i.e. "players", "game", "matchmaking", etc.
    pub module: Cow<'a, str>,

    /// Unique error ID within the error module.
    pub id: u32,

    /// Error message for the developer.
    pub dev_message: Cow<'a, str>,
}
