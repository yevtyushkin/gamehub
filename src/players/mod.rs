mod error;
pub(crate) mod http;
pub mod jwt_service;
mod player;
mod players_db;
pub mod players_service;
mod sign_in_method;
