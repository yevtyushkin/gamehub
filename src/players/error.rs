use crate::api_error::ApiError;
use axum::Json;
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};

/// Possible players errors.
#[derive(Debug, thiserror::Error)]
pub enum PlayersError {
    /// Player not found.
    #[error("player not found")]
    PlayerNotFound,

    /// Error when verifying third party id token.
    #[error("error when verifying third party id token: {0}")]
    IdToken(#[from] id_token_verifier::IdTokenVerifierError),

    /// Error when verifying auth token token.
    #[error("error when verifying auth token: {0}")]
    AuthToken(jsonwebtoken::errors::Error),

    /// When auth token is missing.
    #[error("auth token is missing")]
    AuthTokenMissing,

    /// Internal error.
    #[error("internal error: {0}")]
    Internal(#[from] anyhow::Error),
}

impl IntoResponse for PlayersError {
    fn into_response(self) -> Response {
        let (status, id) = match &self {
            PlayersError::IdToken(_) => (StatusCode::BAD_REQUEST, 0),
            PlayersError::PlayerNotFound => (StatusCode::UNAUTHORIZED, 1),
            PlayersError::AuthToken(_) => (StatusCode::UNAUTHORIZED, 2),
            PlayersError::AuthTokenMissing => (StatusCode::UNAUTHORIZED, 3),
            PlayersError::Internal(_) => (StatusCode::INTERNAL_SERVER_ERROR, 4),
        };

        let body = ApiError {
            module: "players".into(),
            id,
            dev_message: self.to_string().into(),
        };

        (status, Json(body)).into_response()
    }
}
