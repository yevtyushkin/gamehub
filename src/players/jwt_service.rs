use crate::app_state::AppState;
use crate::config::JwtConfig;
use crate::players::error::PlayersError;
use crate::players::player::PlayerId;
use anyhow::Context;
use axum::extract::FromRequestParts;
use axum::http::header::AUTHORIZATION;
use axum::http::request::Parts;
use chrono::{Duration, Utc};
use serde::{Deserialize, Serialize};
use std::borrow::Cow;

/// Operations with [Player]s JWT tokens.
#[cfg_attr(test, mockall::automock)]
pub trait JwtService {
    /// Creates an [AuthToken] from the given [PlayerId].
    fn create_token(&self, player_id: PlayerId) -> Result<AuthToken<'static>, PlayersError>;

    /// Verifies the given [AuthToken] and returns its [AuthTokenClaims].
    #[allow(clippy::needless_lifetimes)]
    fn verify_token<'a>(&self, token: &AuthToken<'a>) -> Result<AuthTokenClaims, PlayersError>;
}

/// Auth token representing a result of a successful sign in.
#[derive(Debug, Clone, Serialize, Deserialize, derive_more::AsRef, PartialEq)]
pub struct AuthToken<'a>(pub Cow<'a, str>);

impl AuthToken<'_> {
    #[cfg(test)]
    /// Returns a test [AuthToken].
    pub fn test() -> AuthToken<'static> {
        AuthToken(Cow::Owned("auth_token".into()))
    }
}

/// [AuthToken] claims shape.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthTokenClaims {
    /// When the token expires.
    pub exp: i64,

    /// When the token was issued.
    pub iat: i64,

    /// Subject [PlayerId] the token is issued for.
    pub sub: PlayerId,
}

impl AuthTokenClaims {
    #[cfg(test)]
    /// Returns a test [AuthTokenClaims].
    pub fn test() -> AuthTokenClaims {
        AuthTokenClaims {
            exp: 123,
            iat: 456,
            sub: PlayerId::test(),
        }
    }
}

impl<S: AppState> FromRequestParts<S> for AuthTokenClaims {
    type Rejection = PlayersError;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        match parts
            .headers
            .get(AUTHORIZATION)
            .and_then(|header_value| header_value.to_str().ok())
        {
            Some(header_value_str) => {
                let jwt_token = AuthToken(Cow::Borrowed(&header_value_str["Bearer ".len()..]));
                let claims = state.jwt_service().verify_token(&jwt_token)?;
                Ok(claims)
            }

            None => Err(PlayersError::AuthTokenMissing),
        }
    }
}

/// Default implementation of [JwtService].
#[derive(Clone)]
pub struct JwtServiceDefault {
    /// JWT [Validation] settings.
    validation: jsonwebtoken::Validation,

    /// JWT [DecodingKey] for signature verification.
    decoding_key: jsonwebtoken::DecodingKey,

    /// JWT [EncodingKey] for signing.
    encoding_key: jsonwebtoken::EncodingKey,

    /// JWT [Header] for JWT creation.
    header: jsonwebtoken::Header,

    /// [Duration] how long the token is valid for.
    token_ttl: Duration,
}

impl JwtServiceDefault {
    /// Creates a new [JwtServiceDefault] with the given [JwtConfig].
    pub fn new(config: JwtConfig) -> JwtServiceDefault {
        let validation = jsonwebtoken::Validation::default();
        let decoding_key = jsonwebtoken::DecodingKey::from_secret(config.secret.as_ref());
        let encoding_key = jsonwebtoken::EncodingKey::from_secret(config.secret.as_ref());
        let header = jsonwebtoken::Header::default();
        let token_ttl = config.ttl;

        JwtServiceDefault {
            validation,
            decoding_key,
            encoding_key,
            header,
            token_ttl,
        }
    }

    #[cfg(test)]
    /// Returns a test [JwtServiceDefault].
    pub fn test() -> JwtServiceDefault {
        JwtServiceDefault::new(JwtConfig::test())
    }
}

impl JwtService for JwtServiceDefault {
    fn create_token(&self, player_id: PlayerId) -> Result<AuthToken<'static>, PlayersError> {
        let now = Utc::now();

        let claims = AuthTokenClaims {
            exp: (now + self.token_ttl).timestamp(),
            iat: now.timestamp(),
            sub: player_id,
        };

        let token = jsonwebtoken::encode(&self.header, &claims, &self.encoding_key)
            .context("create auth token")?;

        Ok(AuthToken(Cow::Owned(token)))
    }

    fn verify_token(&self, token: &AuthToken) -> Result<AuthTokenClaims, PlayersError> {
        let claims = jsonwebtoken::decode(token.as_ref(), &self.decoding_key, &self.validation)
            .map_err(PlayersError::AuthToken)?
            .claims;

        Ok(claims)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn auth_token_claims_json_snapshot() {
        insta::assert_json_snapshot!(&AuthTokenClaims::test());
    }

    #[test]
    fn jwt_service_encode_and_decode_succeeds() -> anyhow::Result<()> {
        let player_id = PlayerId::test();
        let service = JwtServiceDefault::test();
        let ttl = service.token_ttl;

        let token = service.create_token(player_id)?;
        let claims = service.verify_token(&token)?;

        assert_eq!(claims.sub, player_id);
        assert!(claims.iat <= Utc::now().timestamp());
        assert_eq!(claims.iat + ttl.num_seconds(), claims.exp);

        Ok(())
    }

    #[test]
    fn jwt_service_decode_rejects_expired_tokens() -> anyhow::Result<()> {
        let service = JwtServiceDefault::test();
        let now = Utc::now();
        let exp_in_past = now.timestamp() - service.validation.leeway as i64 - 10;
        let iat_in_past = exp_in_past - service.token_ttl.num_seconds();
        let claims_expired = AuthToken(
            jsonwebtoken::encode(
                &service.header,
                &AuthTokenClaims {
                    exp: exp_in_past,
                    iat: iat_in_past,
                    sub: PlayerId::test(),
                },
                &service.encoding_key,
            )?
            .into(),
        );

        let result = service.verify_token(&claims_expired);

        assert!(
            matches!(result, Err(PlayersError::AuthToken(e)) if *e.kind() == jsonwebtoken::errors::ErrorKind::ExpiredSignature)
        );

        Ok(())
    }

    #[test]
    fn jwt_service_decode_rejects_wrong_signatures() -> anyhow::Result<()> {
        let service = JwtServiceDefault::test();
        let service_with_different_secret = {
            let mut service = service.clone();
            service.encoding_key =
                jsonwebtoken::EncodingKey::from_secret("wrong-secret".as_bytes());
            service
        };

        let claims_wrong_signature = service_with_different_secret.create_token(PlayerId::test())?;

        let result = service.verify_token(&claims_wrong_signature);
        assert!(
            matches!(result, Err(PlayersError::AuthToken(e)) if *e.kind() == jsonwebtoken::errors::ErrorKind::InvalidSignature)
        );

        Ok(())
    }
}
