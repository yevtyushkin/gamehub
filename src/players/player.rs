use crate::app_state::AppState;
use crate::players::error::PlayersError;
use crate::players::jwt_service::AuthTokenClaims;
use crate::players::players_service::PlayersService;
use axum::RequestPartsExt;
use axum::extract::FromRequestParts;
use axum::http::request::Parts;
use chrono::{DateTime, Utc};
use petname::{Generator, Petnames};
use serde::*;
use std::str::FromStr;
use uuid::Uuid;

/// Player representation in the application.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Player {
    /// [Player]'s ID.
    pub id: PlayerId,

    /// [Player]'s screen name.
    pub screen_name: PlayerScreenName,

    /// When a [Player] has joined.
    pub joined_at: PlayerJoinedAt,
}

impl Player {
    #[cfg(test)]
    /// Returns a test [Player].
    pub fn test() -> Player {
        Player {
            id: PlayerId::test(),
            screen_name: PlayerScreenName::test(),
            joined_at: PlayerJoinedAt::test(),
        }
    }
}

impl<S: AppState> FromRequestParts<S> for Player {
    type Rejection = PlayersError;
    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let claims: AuthTokenClaims = parts.extract_with_state(state).await?;
        let player = state.players_service().player_by_id(&claims.sub).await?;
        Ok(player)
    }
}

/// [Player]'s ID.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq, sqlx::Type)]
pub struct PlayerId(pub Uuid);

impl PlayerId {
    /// Creates a new random [PlayerId].
    pub fn random() -> PlayerId {
        PlayerId(Uuid::now_v7())
    }

    #[cfg(test)]
    /// Returns a test [PlayerId].
    pub fn test() -> PlayerId {
        PlayerId(Uuid::from_u128(1234567890))
    }
}

/// [Player]'s screen name.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, sqlx::Type)]
pub struct PlayerScreenName(String);

impl PlayerScreenName {
    /// Creates a new random [PlayerScreenName].
    pub fn random() -> PlayerScreenName {
        let value = Petnames::small()
            .generate_one(3, "-")
            .unwrap_or_else(|| "no-screen-name".into());

        PlayerScreenName(value)
    }

    #[cfg(test)]
    /// Returns a test [PlayerScreenName].
    pub fn test() -> PlayerScreenName {
        PlayerScreenName("test-screen-name".into())
    }

    /// The maximum size of a [PlayerScreenName] in bytes.
    pub const MAX_SIZE: usize = 30;
}

/// [Player]'s screen name validation error.
#[derive(Debug, Clone, Copy, Eq, PartialEq, thiserror::Error)]
pub enum InvalidPlayerScreenName {
    /// [PlayerScreenName] is empty.
    #[error("player screen name is empty")]
    Empty,

    /// [PlayerScreenName] exceeds [PlayerScreenName::MAX_SIZE].
    #[error(
        "player screen name is larger than allowed max size of {} bytes",
        PlayerScreenName::MAX_SIZE
    )]
    ExceedsMaxSize,
}

impl FromStr for PlayerScreenName {
    type Err = InvalidPlayerScreenName;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.trim();

        if s.is_empty() {
            Err(InvalidPlayerScreenName::Empty)?
        }
        if s.len() > PlayerScreenName::MAX_SIZE {
            Err(InvalidPlayerScreenName::ExceedsMaxSize)?
        }

        Ok(PlayerScreenName(s.into()))
    }
}

/// When the [Player] joined.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, sqlx::Type, PartialEq)]
pub struct PlayerJoinedAt(pub DateTime<Utc>);

impl PlayerJoinedAt {
    /// Creates a new [PlayerJoinedAt] with the current time.
    pub fn now() -> PlayerJoinedAt {
        PlayerJoinedAt(Utc::now())
    }

    #[cfg(test)]
    /// Returns a test [PlayerJoinedAt].
    pub fn test() -> PlayerJoinedAt {
        PlayerJoinedAt(DateTime::UNIX_EPOCH)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn player_json_snapshot() {
        insta::assert_json_snapshot!(&Player::test());
    }

    #[test]
    fn player_screen_name_from_str_succeeds_if_screen_name_is_valid_of_min_size() {
        assert_eq!(
            PlayerScreenName::from_str("a"),
            Ok(PlayerScreenName("a".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str(" a  "),
            Ok(PlayerScreenName("a".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("г"),
            Ok(PlayerScreenName("г".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("  г     "),
            Ok(PlayerScreenName("г".into()))
        );
    }

    #[test]
    fn player_screen_name_from_str_succeeds_if_screen_name_is_valid_of_max_size() {
        let value = "w".repeat(PlayerScreenName::MAX_SIZE);
        assert_eq!(
            PlayerScreenName::from_str(&value),
            Ok(PlayerScreenName(value.clone()))
        );

        let value_ws = format!("  {value}   ");
        assert_eq!(
            PlayerScreenName::from_str(&value_ws),
            Ok(PlayerScreenName(value))
        );

        let value = "г".repeat(PlayerScreenName::MAX_SIZE / 2);
        assert_eq!(
            PlayerScreenName::from_str(&value),
            Ok(PlayerScreenName(value.clone()))
        );

        let value_ws = format!("  {value}   ");
        assert_eq!(
            PlayerScreenName::from_str(&value_ws),
            Ok(PlayerScreenName(value))
        );
    }

    #[test]
    fn player_screen_name_from_str_succeeds_if_screen_name_is_valid_of_medium_size() {
        assert_eq!(
            PlayerScreenName::from_str("abcd1234"),
            Ok(PlayerScreenName("abcd1234".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("  abcd1234     "),
            Ok(PlayerScreenName("abcd1234".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("абв123"),
            Ok(PlayerScreenName("абв123".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("  абв123     "),
            Ok(PlayerScreenName("абв123".into()))
        );
    }

    #[test]
    fn player_screen_name_from_str_fails_if_screen_name_is_empty() {
        assert_eq!(
            PlayerScreenName::from_str(""),
            Err(InvalidPlayerScreenName::Empty)
        );

        assert_eq!(
            PlayerScreenName::from_str("     "),
            Err(InvalidPlayerScreenName::Empty)
        );
    }

    #[test]
    fn player_screen_name_from_str_fails_if_screen_name_exceeds_max_size() {
        let value = "w".repeat(PlayerScreenName::MAX_SIZE + 1);
        assert_eq!(
            PlayerScreenName::from_str(&value),
            Err(InvalidPlayerScreenName::ExceedsMaxSize)
        );

        let value_ws = format!("  {value}   ");
        assert_eq!(
            PlayerScreenName::from_str(&value_ws),
            Err(InvalidPlayerScreenName::ExceedsMaxSize)
        );

        let value = "г".repeat(PlayerScreenName::MAX_SIZE / 2 + 1);
        assert_eq!(
            PlayerScreenName::from_str(&value),
            Err(InvalidPlayerScreenName::ExceedsMaxSize)
        );

        let value_ws = format!("  {value}   ");
        assert_eq!(
            PlayerScreenName::from_str(&value_ws),
            Err(InvalidPlayerScreenName::ExceedsMaxSize)
        );
    }
}
