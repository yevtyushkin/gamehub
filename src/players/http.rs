use crate::app_state::AppState;
use crate::players::error::PlayersError;
use crate::players::jwt_service::AuthToken;
use crate::players::player::Player;
use crate::players::players_service::PlayersService;
use axum::extract::State;
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::routing::*;
use axum::{Json, Router};
use serde::{Deserialize, Serialize};

/// A sign-in request.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum SignInRequest {
    /// Sign in with Google.
    Google {
        /// Google issued [IdToken].
        id_token: IdToken,
    },
}

impl SignInRequest {
    #[cfg(test)]
    /// Returns a test [SignInRequest::Google].
    pub fn test_google() -> SignInRequest {
        SignInRequest::Google {
            id_token: IdToken::test(),
        }
    }
}

/// Response to a [SignInRequest] in case of success.
#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct SignInResponse<'a> {
    /// [AuthToken] to use in subsequent requests.
    pub auth_token: AuthToken<'a>,
}

impl SignInResponse<'_> {
    #[cfg(test)]
    /// Returns a test [SignInResponse].
    pub fn test() -> SignInResponse<'static> {
        SignInResponse {
            auth_token: AuthToken::test(),
        }
    }
}

/// ID token containing user information.
#[derive(
    Debug, Clone, Deserialize, Serialize, derive_more::AsRef, derive_more::Deref, PartialEq,
)]
pub struct IdToken(pub String);

impl IdToken {
    #[cfg(test)]
    /// Returns a test [IdToken].
    pub fn test() -> IdToken {
        IdToken("test-id-token".into())
    }
}

/// [Router] for the [crate::players] module.
pub fn router<S: AppState>() -> Router<S> {
    Router::new().nest(
        "/players",
        Router::new()
            .route("/sign_in", post(sign_in::<S>))
            .route("/player_info", get(player_info)),
    )
}

/// `/sign_in` handler. Handles [SignInRequest] and returns [SignInResponse] in case of success.
async fn sign_in<S: AppState>(
    State(app_state): State<S>,
    Json(request): Json<SignInRequest>,
) -> Result<Response, PlayersError> {
    let auth_token = app_state.players_service().sign_in(&request).await?;

    let body = SignInResponse { auth_token };
    let response = (StatusCode::OK, Json(body)).into_response();

    Ok(response)
}

/// `/player_info` handler. Returns current [Player] information.
async fn player_info(player: Player) -> Json<Player> {
    Json(player)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::api_error::ApiError;
    use crate::app_state::MockAppState;
    use crate::players::jwt_service::{AuthTokenClaims, MockJwtService};
    use crate::players::player::PlayerId;
    use crate::players::players_service::MockPlayersService;
    use axum_test::TestServer;
    use axum_test::http::header::AUTHORIZATION;
    use id_token_verifier::IdTokenVerifierError;
    use id_token_verifier::validation::ValidationError;
    use jsonwebtoken::errors::ErrorKind;
    use mockall::predicate::eq;
    use std::sync::Arc;

    #[test]
    fn sign_in_request_json_snapshot() {
        insta::assert_json_snapshot!(&SignInRequest::test_google());
    }

    #[test]
    fn sign_in_response_json_snapshot() {
        insta::assert_json_snapshot!(&SignInResponse::test());
    }

    #[tokio::test]
    async fn sign_in_handler_returns_correct_response_when_players_service_succeds()
    -> anyhow::Result<()> {
        let mut players_service = MockPlayersService::new();
        players_service
            .expect_sign_in()
            .with(eq(SignInRequest::test_google()))
            .returning(|_| Box::pin(async { Ok(AuthToken::test()) }));
        let state = Arc::new(MockAppState::default().with_players_service(players_service));
        let server = TestServer::new(router().with_state(state))?;

        let response = server
            .post("/players/sign_in")
            .json(&SignInRequest::test_google())
            .await;

        response.assert_status(StatusCode::OK);
        response.assert_json(&SignInResponse::test());

        Ok(())
    }

    #[tokio::test]
    async fn sign_in_handler_returns_correct_response_when_players_service_fails()
    -> anyhow::Result<()> {
        let mut players_service = MockPlayersService::new();
        players_service
            .expect_sign_in()
            .with(eq(SignInRequest::test_google()))
            .returning(|_| {
                Box::pin(async {
                    Err(PlayersError::IdToken(IdTokenVerifierError::Validation(
                        ValidationError::MissingKeyId,
                    )))
                })
            });
        let state = Arc::new(MockAppState::default().with_players_service(players_service));
        let server = TestServer::new(router().with_state(state))?;

        let response = server
            .post("/players/sign_in")
            .json(&SignInRequest::test_google())
            .await;

        response.assert_status(StatusCode::BAD_REQUEST);
        let error = response.json::<ApiError>();
        assert_eq!(error.module, "players");
        assert_eq!(error.id, 0);

        Ok(())
    }

    #[tokio::test]
    async fn player_info_handler_fails_if_auth_token_missing() -> anyhow::Result<()> {
        let state = Arc::new(MockAppState::default());
        let server = TestServer::new(router().with_state(state))?;

        let response = server.get("/players/player_info").await;

        response.assert_status(StatusCode::UNAUTHORIZED);
        let error = response.json::<ApiError>();
        assert_eq!(error.module, "players");
        assert_eq!(error.id, 3);
        assert_eq!(error.dev_message, "auth token is missing");

        Ok(())
    }

    #[tokio::test]
    async fn player_info_handler_fails_if_auth_token_validation_fails() -> anyhow::Result<()> {
        let mut jwt_service = MockJwtService::new();
        jwt_service
            .expect_verify_token()
            .withf(|token| token.as_ref() == "invalid")
            .returning(|_| {
                Err(PlayersError::AuthToken(jsonwebtoken::errors::Error::from(
                    ErrorKind::InvalidToken,
                )))
            });

        let state = Arc::new(MockAppState::default().with_jwt_service(jwt_service));
        let mut server = TestServer::new(router().with_state(state))?;
        server.add_header(AUTHORIZATION, "Bearer invalid");

        let response = server.get("/players/player_info").await;

        response.assert_status(StatusCode::UNAUTHORIZED);
        let error = response.json::<ApiError>();
        assert_eq!(error.module, "players");
        assert_eq!(error.id, 2);

        Ok(())
    }

    #[tokio::test]
    async fn player_info_handler_fails_if_player_lookup_fails() -> anyhow::Result<()> {
        let mut jwt_service = MockJwtService::new();
        jwt_service
            .expect_verify_token()
            .withf(|token| token.as_ref() == "valid")
            .returning(|_| Ok(AuthTokenClaims::test()));

        let mut players_service = MockPlayersService::new();
        players_service
            .expect_player_by_id()
            .withf(|player_id| player_id == &PlayerId::test())
            .returning(|_| Box::pin(async { Err(PlayersError::PlayerNotFound) }));

        let state = Arc::new(
            MockAppState::default()
                .with_jwt_service(jwt_service)
                .with_players_service(players_service),
        );
        let mut server = TestServer::new(router().with_state(state))?;
        server.add_header(AUTHORIZATION, "Bearer valid");

        let response = server.get("/players/player_info").await;

        response.assert_status(StatusCode::UNAUTHORIZED);
        let error = response.json::<ApiError>();
        assert_eq!(error.module, "players");
        assert_eq!(error.id, 1);

        Ok(())
    }
}
