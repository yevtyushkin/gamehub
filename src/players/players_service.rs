use crate::players::error::*;
use crate::players::http::*;
use crate::players::jwt_service::*;
use crate::players::player::*;
use crate::players::players_db::*;
use crate::players::sign_in_method::*;
use id_token_verifier::*;
use serde::*;
use sqlx::PgPool;

/// Provides logic working with [Player]s.
#[cfg_attr(test, mockall::automock)]
pub trait PlayersService {
    /// Handles the given [SignInRequest].
    fn sign_in(
        &self,
        request: &SignInRequest,
    ) -> impl Future<Output = Result<AuthToken<'static>, PlayersError>> + Send;

    /// Returns a [Player] by the given [PlayerId].
    fn player_by_id(
        &self,
        player_id: &PlayerId,
    ) -> impl Future<Output = Result<Player, PlayersError>> + Send;
}

/// Default [PlayersService] implementation.
pub struct PlayersServiceDefault<D = PgPool, GV = IdTokenVerifierDefault, JS = JwtServiceDefault> {
    /// [PlayersDb] for [Player]s db operations.
    players_db: D,

    /// [IdTokenVerifier] implementation for [ThirdPartySignInProvider::Google] ID tokens.
    google_id_token_verifier: GV,

    /// [JwtService] implementation to work with [Player]s JWT tokens.
    jwt_service: JS,
}

impl PlayersServiceDefault {
    /// Creates a new [PlayersServiceDefault] with the given [PgPool], Google
    /// [IdTokenVerifierConfig] and [JwtServiceDefault].
    pub fn new(
        pg_pool: PgPool,
        http_client: reqwest::Client,
        google_id_token_verifier_config: IdTokenVerifierConfig,
        jwt_service: JwtServiceDefault,
    ) -> PlayersServiceDefault {
        let players_db = pg_pool;
        let google_id_token_verifier =
            IdTokenVerifierDefault::new(google_id_token_verifier_config, http_client);

        PlayersServiceDefault {
            players_db,
            google_id_token_verifier,
            jwt_service,
        }
    }
}

impl<D, GV, JS> PlayersService for PlayersServiceDefault<D, GV, JS>
where
    D: PlayersDb + Sync,
    GV: IdTokenVerifier + Sync,
    JS: JwtService + Sync,
{
    async fn sign_in(&self, request: &SignInRequest) -> Result<AuthToken<'static>, PlayersError> {
        let sign_in_method = match request {
            SignInRequest::Google { id_token } => {
                let claims = self
                    .google_id_token_verifier
                    .verify::<ThirdPartyIdTokenClaims>(id_token.as_ref())
                    .await?;

                SignInMethod::ThirdParty(ThirdPartySignInMethod {
                    provider: ThirdPartySignInProvider::Google,
                    user_id: claims.sub,
                })
            }
        };

        let player = match self
            .players_db
            .find_player_with_sign_in_method(&sign_in_method)
            .await
        {
            Ok(player) => player,
            Err(PlayersError::PlayerNotFound) => {
                let player = Player {
                    id: PlayerId::random(),
                    screen_name: PlayerScreenName::random(),
                    joined_at: PlayerJoinedAt::now(),
                };

                self.players_db
                    .create_player_with_sign_in_method(&player, &sign_in_method)
                    .await?;

                player
            }
            e => e?,
        };

        let auth_token = self.jwt_service.create_token(player.id)?;

        Ok(auth_token)
    }

    async fn player_by_id(&self, player_id: &PlayerId) -> Result<Player, PlayersError> {
        self.players_db.find_player_by_id(player_id).await
    }
}

/// Target ID token claims. Used with [IdTokenVerifier::verify] when signing in with third party
/// sign in providers.
#[derive(Deserialize, Serialize)]
struct ThirdPartyIdTokenClaims {
    /// Subject [ThirdPartySignInUserId] the ID token is issued for.
    sub: ThirdPartySignInUserId,
}

impl ThirdPartyIdTokenClaims {
    #[cfg(test)]
    /// Returns a test [ThirdPartyIdTokenClaims].
    fn test() -> ThirdPartyIdTokenClaims {
        ThirdPartyIdTokenClaims {
            sub: ThirdPartySignInUserId::test(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use id_token_verifier::validation::ValidationError;
    use mockall::mock;
    use mockall::predicate::eq;
    use serde::de::DeserializeOwned;

    mock! {
        IdTokenVerifier {
            fn verify<T: DeserializeOwned + 'static>(
                &self,
                id_token: &str
            ) -> impl Future<Output = Result<T, IdTokenVerifierError>> + Send;
        }
    }
    impl IdTokenVerifier for MockIdTokenVerifier {
        fn verify<T: DeserializeOwned + 'static>(
            &self,
            id_token: &str,
        ) -> impl Future<Output = Result<T, IdTokenVerifierError>> + Send {
            self.verify(id_token)
        }
    }

    #[test]
    fn third_party_id_token_claims_json_snapshot() {
        insta::assert_json_snapshot!(&ThirdPartyIdTokenClaims::test());
    }

    #[tokio::test]
    async fn players_service_sign_in_fails_if_google_id_token_verification_fails() {
        let mut google_id_token_verifier = MockIdTokenVerifier::new();
        google_id_token_verifier
            .expect_verify::<ThirdPartyIdTokenClaims>()
            .withf(|id_token| id_token == IdToken::test().as_str())
            .returning(|_| {
                Box::pin(async {
                    Err(IdTokenVerifierError::Validation(
                        ValidationError::MissingKeyId,
                    ))
                })
            });

        let service = PlayersServiceDefault {
            google_id_token_verifier,
            players_db: MockPlayersDb::new(),
            jwt_service: MockJwtService::new(),
        };

        let result = service.sign_in(&SignInRequest::test_google()).await;

        assert!(matches!(
            result,
            Err(PlayersError::IdToken(IdTokenVerifierError::Validation(
                ValidationError::MissingKeyId,
            )))
        ));
    }

    #[tokio::test]
    async fn players_service_sign_in_fails_if_db_player_lookup_fails() {
        let mut google_id_token_verifier = MockIdTokenVerifier::new();
        google_id_token_verifier
            .expect_verify::<ThirdPartyIdTokenClaims>()
            .withf(|id_token| id_token == IdToken::test().as_str())
            .returning(|_| Box::pin(async { Ok(ThirdPartyIdTokenClaims::test()) }));

        let mut players_db = MockPlayersDb::new();
        players_db
            .expect_find_player_with_sign_in_method()
            .with(eq(SignInMethod::test_google()))
            .returning(|_| {
                Box::pin(async { Err(PlayersError::Internal(anyhow::anyhow!("oops"))) })
            });

        let service = PlayersServiceDefault {
            players_db,
            google_id_token_verifier,
            jwt_service: MockJwtService::new(),
        };

        let result = service.sign_in(&SignInRequest::test_google()).await;

        assert!(matches!(result, Err(PlayersError::Internal(e)) if e.to_string() == "oops"));
    }

    #[tokio::test]
    async fn players_service_sign_in_creates_player_if_player_does_not_exist() {
        let mut google_id_token_verifier = MockIdTokenVerifier::new();
        google_id_token_verifier
            .expect_verify::<ThirdPartyIdTokenClaims>()
            .withf(|id_token| id_token == IdToken::test().as_str())
            .returning(|_| Box::pin(async { Ok(ThirdPartyIdTokenClaims::test()) }));

        let mut players_db = MockPlayersDb::new();
        players_db
            .expect_find_player_with_sign_in_method()
            .with(eq(SignInMethod::test_google()))
            .returning(|_| Box::pin(async { Err(PlayersError::PlayerNotFound) }));
        players_db
            .expect_create_player_with_sign_in_method()
            .with(
                mockall::predicate::always(),
                eq(SignInMethod::test_google()),
            )
            .returning(|_, _| Box::pin(async { Ok(()) }));

        let mut jwt_service = MockJwtService::new();
        jwt_service
            .expect_create_token()
            .returning(|_| Ok(AuthToken::test()));

        let service = PlayersServiceDefault {
            players_db,
            google_id_token_verifier,
            jwt_service,
        };

        let auth_token = service
            .sign_in(&SignInRequest::test_google())
            .await
            .unwrap();

        assert_eq!(auth_token, AuthToken::test());
    }

    #[tokio::test]
    async fn players_service_sign_in_creates_auth_token_from_existing_player() {
        let mut google_id_token_verifier = MockIdTokenVerifier::new();
        google_id_token_verifier
            .expect_verify::<ThirdPartyIdTokenClaims>()
            .withf(|id_token| id_token == IdToken::test().as_str())
            .returning(|_| Box::pin(async { Ok(ThirdPartyIdTokenClaims::test()) }));

        let mut players_db = MockPlayersDb::new();
        players_db
            .expect_find_player_with_sign_in_method()
            .with(eq(SignInMethod::test_google()))
            .returning(|_| Box::pin(async { Ok(Player::test()) }));

        let mut jwt_service = MockJwtService::new();
        jwt_service
            .expect_create_token()
            .with(eq(PlayerId::test()))
            .returning(|_| Ok(AuthToken::test()));

        let service = PlayersServiceDefault {
            players_db,
            google_id_token_verifier,
            jwt_service,
        };

        let auth_token = service
            .sign_in(&SignInRequest::test_google())
            .await
            .unwrap();

        assert_eq!(auth_token, AuthToken::test());
    }

    #[tokio::test]
    async fn players_service_player_by_id_lookups_player_in_db() {
        let mut players_db = MockPlayersDb::new();
        players_db
            .expect_find_player_by_id()
            .with(eq(PlayerId::test()))
            .returning(|_| Box::pin(async { Ok(Player::test()) }));

        let service = PlayersServiceDefault {
            players_db,
            google_id_token_verifier: MockIdTokenVerifier::new(),
            jwt_service: MockJwtService::new(),
        };

        let player = service.player_by_id(&PlayerId::test()).await.unwrap();

        assert_eq!(player, Player::test());
    }
}
