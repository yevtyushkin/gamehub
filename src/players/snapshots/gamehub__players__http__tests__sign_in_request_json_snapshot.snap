---
source: src/players/http.rs
expression: "&SignInRequest::test_google()"
---
{
  "Google": {
    "id_token": "test-id-token"
  }
}
